enclave/core/switchlesscalls.c

```diff
@@ -44,6 +51,16 @@ oe_result_t oe_handle_init_switchless(uint64_t arg_in)
     oe_switchless_call_manager_t safe_manager;
     size_t contexts_size, threads_size;
 
+    // Check if already initialized
+    if (!oe_atomic_compare_and_swap(
```

Comment:
Nice work using atomics instead of locking.

When two threads simultaneously call oe_handle_init_switchless

- one of them will atomically set is_switchless_initialized flag to true **and then** initialize the data structures.
- the other thread will see that the flag is true and error out

Another scenario to solve is 
- When one thread has set the flag to true, but **not yet completed initialization**
- Another enclave thread could check the flag (via oe_is_switchless_initialized) and incorrectly think that initialization is complete and go on to make a switchless call.

Reply:
Good point. I have updated the ordering of the checks and added a comment for the reasoning.

Reply:
@jhand2 Your comments are related for the 1st scenario only. 

For the 2nd scenario, depending on how much progress the initializing thread has made after setting `_is_switchless_initialized` to true. The 2nd thread may either:

* fall back to regular OCALL if the first thread has not set `_host_worker_count`
* crash if the first thread has set `_host_worker_count`, but not `_host_worker_contexts`

Ideally, we should make `oe_is_switchless_initialized` checking `_host_worker_contexts` instead of a flag doesn't really indicate the initialization has been completed or not.

Reply:
The only way to protect against that is to use a lock. Individual atomic operations are only effective if the variables can be updated independently. But if you want to update all 3 variables in one atomic operation I think a spin lock is required here.

cc @anakrish

Reply:
Plus we would be locking a very small section of code so the perf hit of using a spinlock is very minimal

Reply:
There are a few ways to implement this correctly:

- Use a spinlock in both `oe_handle_init_switchless` and `oe_is_switchless_initialized`. Make sure the spinlock is released in error scenarios.

- Use one atomic-flag to make sure `oe_handle_init_switchless` performs initialization only once. Use another atomic-flag to check if initialization is complete or not. This second flag must follow acquire-release semantics or the compiler could reorder instructions and set this flag at a different location than written in code. This is in essense what Xuejun is suggesting.

There is also `oe_once` which allows initialization functions to be called only once. However, it does not support initialization functions with arguments.

Reply:
IMO spinlock is cleaner and probably more efficient (although I'm not that worried about perf in this case).

Using two atomically updated flags is a little hard to follow from a readability perspective.

Anyone have another opinion here?

Reply:
@anakrish I took a shot at the second one. Can you take a look?